package GatkSNPCalling;

import java.io.*;

public class SnpReadCount {
    private String refGenomeFilePath, bamFile, gatk;
    private File gatkSnpFile, outputSnpReadFile;

    /**
     *
     * @param refGenomeFilePath reference genome fasta file
     * @param bamFile bam file generate by STAR
     * @param gatk gatk executive file
     * @param gatkSnpFile SNP sites which are called by GATK
     */
    public SnpReadCount(String refGenomeFilePath, String bamFile, String gatk, String gatkSnpFile) {
        this.refGenomeFilePath = refGenomeFilePath;
        this.bamFile = bamFile;
        this.gatk = gatk;
        this.gatkSnpFile = new File(gatkSnpFile);
        this.outputSnpReadFile = new File(bamFile.substring(0, bamFile.lastIndexOf("_"))+"ReadsCount.txt");
    }

    /**
     * use gatk ASEReadCounter get reads count for each RNA-Seq snp, the input file is sorted bam file which is generated by STAR
     */
    public void countRead() {
        String cmd = this.gatk + " ASEReadCounter -I " + this.bamFile + " -V " + this.gatkSnpFile + " -R " + this.refGenomeFilePath
                + " --output " + this.outputSnpReadFile;
        
        try {
            Process p =  Runtime.getRuntime().exec(cmd);
            int exitVal = p.waitFor();
            if (exitVal != 0) {
                System.out.println("failed to get ASE reads count");
                System.exit(2);
            }
        } catch (IOException | InterruptedException ie) {
            ie.printStackTrace();
            System.exit(2);
        }
    }
}

